interfaces/apparmor/spec.go

```diff
@@ -122,6 +136,84 @@ func (spec *Specification) AddDeduplicatedSnippet(snippet string) {
 	}
 }
 
+// AddParametricSnippet adds a new apparmor snippet both de-duplicated and optimized for the parser.
+//
+// Conceptually the function takes a parametric template and a single value to
+// remember. The resulting snippet text is a single entry resulting from the
+// expanding the template and all the unique values observed, in the order they
+// were observed.
+//
+// The template is expressed as a slice of strings, with the parameter
+// automatically injected between any two of them, or in the special case of
+// only one fragment, after that fragment.
+//
+// The resulting expansion depends on the number of values seen. If only one
+// value is seen the resulting snippet is just the plain string one would
+// expect if no parametric optimization had taken place. If more than one
+// distinct value was seen then the resulting apparmor rule uses alternation
+// syntax {param1,param2,...,paramN} which has better compilation time and
+// memory complexity as compared to a set of naive expansions of the full
+// snippet one after another.
+//
+// For example the code:
+//
+// 		AddParametricSnippet([]{"/dev/", "rw,", "sda1")
+//		AddParametricSnippet([]{"/dev/", "rw,", "sda3")
+//		AddParametricSnippet([]{"/dev/", "rw,", "sdb2")
+//
+// Results in a single apparmor rule:
+//
+//		"/dev/{sda1,sda3,sdb2} rw,"
+//
+// It should be used whenever the apparmor template combines more than one use
+// of "**" syntax (which represent arbitrary many directories or files) and a
+// variable component, like a device name or similar.
+func (spec *Specification) AddParametricSnippet(templateFragment []string, value string) {
+	if len(spec.securityTags) == 0 {
+		return
+	}
+
+	// We need to build a template string from the templateFragment.
+	//
+	// If only a single fragment is given we just  append our "###PARM###":
+	//  []string{"prefix"} becomes -> "prefix###PARAM###"
+	//
+	// Otherwise we join the strings:
+	//  []string{"pre","post"} becomes -> "pre###PARAM###post"
+	//
+	// This seems to be the most natural way of doing this.
+	var template string
+	switch len(templateFragment) {
+	case 0:
+		return
+	case 1:
+		template = templateFragment[0] + "###PARAM###"
+	default:
+		template = strings.Join(templateFragment, "###PARAM###")
+	}
```

Comment:
The code comment only specifies templateFragment as a length of two, I wonder if we should enforce that where we have:
```
	switch len(templateFragment) {
	case 1:
		template = templateFragment[0] + "###PARAM###"
	case 2:
		template = strings.Join(templateFragment, "###PARAM###")
	default:
		return
	}
```
Not a blocker; I was just have trouble coming up with a use case for `[]string{"pre","post","zpost"} -> "pre###PARAM###post###PARAM###zpost"`

Reply:
Bikeshed Friday.

Maybe we should panic or raise some `internal error: ..` when the template length is > 2. I actually found the implicit param addition quite confusion.

Right now we can also call `AddParametricSnippet(nil, "foo")` and it returns silently. Pretty much the same as `AddParametricSnippet("foobar foobar")` (without `###PARAM###`) if we just tried to string.Split() and check the length == 0.

How about this (combines @zyga's idea we discussed on IRC) or some variation:
```
type ParametricSnippetTemplate {
   template []string
}

func MustParametrizeOn(template, param string) ParametricSnippetTemplate {
    tpl := strings.Split(template, param)
    if len(tpl) == 0 || len(tpl) > 2 {
        panic(fmt.Sprintf("invalid template %q, parameter appears too frequently (%d)", template, len(tpl))
    }
    return ParametricSnippetTemplate{ template: tpl }
}

func (spec *Specification) AddParametricSnippet(template ParametricSnippetTemplate, value string) {
   ...
}

// then when using

var devSdaRw = MustParametrizeOn("/dev/sda###PARAM### rw,", "###PARAM###")

func doMagic(...) {
    spec.AddParametricSnippet(devSdaRw, "1")
}
```

Reply:
AIUI, the panic() was removed intentionally. Perhaps it would be fine to have something like this:
```
	switch len(templateFragment) {
	case 1:
		template = templateFragment[0] + "###PARAM###"
	case 2:
		template = strings.Join(templateFragment, "###PARAM###")
	default:
		log error
		return
	}
```

As for the other code changes, I'm not sure they improve readability (iirc, use of '###PARAM###' in the rule was deemed inelegant so it was changed to what we have now). I didn't personally mind '###PARAM###' in the rule since we have a similar pattern with dbus rules for the peer label.

I don't really care for using '###PARAM###' twice as in the newly proposed `var devSdaRw = MustParametrizeOn("/dev/sda###PARAM### rw,", "###PARAM###")`. Assuming the i2c changes I recommended are incorporated, I'll defer to the snapd team on code style though so none of this is a blocker.

Reply:
Note that in my suggestion the panic cannot happen at "late" runtime. It will always panic in tests if the syntax is incorrect.